disco_aws_automation/disco_group.py

```diff
@@ -0,0 +1,99 @@
+"""Contains DiscoGroup class that is above all other group classes used"""
+import logging
+
+from abc import ABCMeta, abstractmethod
+
+from .disco_autoscale import DiscoAutoscale
+from .disco_elastigroup import DiscoElastigroup
+
+logger = logging.getLogger(__name__)
+
+
+class BaseGroup(object):
+    """Abstract class definition for AWS groups"""
+    __metaclass__ = ABCMeta
+
+    @abstractmethod
+    def get_existing_group(self, hostclass, group_name, throw_on_two_groups):
+        """Get list of group objects for a hostclass"""
+        return
+
+    @abstractmethod
+    def get_existing_groups(self, hostclass, group_name):
+        """Get list of all group objects for a hostclass"""
+        return
+
+    @abstractmethod
+    def get_instances(self, hostclass, group_name):
+        """Get list of instances in groups"""
+        return
+
+    @abstractmethod
+    def delete_groups(self, hostclass, group_name, force):
+        """Delete groups of a hostclass"""
+        return
+
+    @abstractmethod
+    def scaledown_groups(self, hostclass, group_name, wait, noerror):
+        """Scale down number of instances in a group"""
+        return
+
+
+class DiscoGroup(BaseGroup):
+    """Implementation of DiscoGroup regardless of the type of group"""
+
+    def __init__(self, environment_name):
+        """Implementation of BaseGroup in AWS"""
+        self.environment_name = environment_name
+        self._autoscale = DiscoAutoscale(environment_name=self.environment_name)
+        self._elastigroup = DiscoElastigroup(environment_name=self.environment_name)
+
+    def get_existing_group(self, hostclass=None, group_name=None, throw_on_two_groups=True):
+        asg_group = self._autoscale.get_existing_group(
+            hostclass=hostclass,
+            group_name=group_name,
+            throw_on_two_groups=throw_on_two_groups
+        )
+        spot_group = self._elastigroup.get_existing_group(
+            hostclass=hostclass,
+            group_name=group_name,
+            throw_on_two_groups=throw_on_two_groups
+        )
+        if asg_group and spot_group:
+            return sorted([asg_group.__dict__, spot_group], key=lambda grp: grp['name'], reverse=True)[0]
```

Comment:
It would help to change `self._autoscale.get_existing_group` to always return a dict so that it matches spot inst. The return type is part of a method's interface so both methods should return the same type of object